src/algebraic_geometry/elliptic_curve/point.lean

```diff
@@ -0,0 +1,500 @@
+lemma nonsingular_add_of_ne' :
+  W.nonsingular (W.add_X x₁ x₂ $ slope_of_ne x₁ x₂ y₁ y₂)
+    (W.add_Y' x₁ x₂ y₁ $ slope_of_ne x₁ x₂ y₁ y₂) :=
+begin
+  by_cases hx₁ : W.add_X x₁ x₂ (slope_of_ne x₁ x₂ y₁ y₂) = x₁,
+  { rwa [add_Y', hx₁, sub_self, mul_zero, zero_add] },
+  { by_cases hx₂ : W.add_X x₁ x₂ (slope_of_ne x₁ x₂ y₁ y₂) = x₂,
+    { rwa [add_Y', ← neg_sub, mul_neg, hx₂, slope_of_ne, div_mul_cancel _ $ sub_ne_zero_of_ne hx,
+           neg_sub, sub_add_cancel] },
+    { apply nonsingular_add_of_eval_derivative_ne_zero,
+      rw [derivative_add_polynomial_of_ne h₁ h₂ hx],
+      eval_simp,
+      simpa only [neg_ne_zero, sub_self, mul_zero, add_zero]
+        using mul_ne_zero (sub_ne_zero_of_ne hx₁) (sub_ne_zero_of_ne hx₂) } }
+end
+
+/-- The addition of two nonsingular points in `W` with distinct $X$-coordinates is nonsingular. -/
+lemma nonsingular_add_of_ne :
+  W.nonsingular (W.add_X x₁ x₂ $ slope_of_ne x₁ x₂ y₁ y₂)
+    (W.add_Y x₁ x₂ y₁ $ slope_of_ne x₁ x₂ y₁ y₂) :=
+nonsingular_neg $ nonsingular_add_of_ne' h₁ h₂ h₁' h₂' hx
+
+omit h₁ h₂ h₁' h₂' hx
+
+namespace point
+
+open_locale classical
+
+/-- The addition of two nonsingular rational points.
+
+Given two nonsingular rational points `P` and `Q`, use `P + Q` instead of `add P Q`. -/
+noncomputable def add : W.point → W.point → W.point
+| 0                          P                          := P
+| P                          0                          := P
+| (@some _ _ _ x₁ y₁ h₁ h₁') (@some _ _ _ x₂ y₂ h₂ h₂') :=
+if hx : x₁ = x₂ then if hy : y₁ = W.neg_Y x₂ y₂ then 0
+else some (equation_add_of_eq h₁ $ Y_ne_of_Y_ne h₁ h₂ hx hy)
+  (nonsingular_add_of_eq h₁ h₁' $ Y_ne_of_Y_ne h₁ h₂ hx hy)
+else some (equation_add_of_ne h₁ h₂ hx) (nonsingular_add_of_ne h₁ h₂ h₁' h₂' hx)
+
+noncomputable instance : has_add W.point := ⟨add⟩
+
+@[simp] lemma add_def (P Q : W.point) : P.add Q = P + Q := rfl
+
+noncomputable instance : add_zero_class W.point :=
+⟨0, (+), by rintro (_ | _); refl, by rintro (_ | _); refl⟩
+
+@[simp] lemma some_add_some_of_y_eq (hx : x₁ = x₂) (hy : y₁ = W.neg_Y x₂ y₂) :
+  some h₁ h₁' + some h₂ h₂' = 0 :=
+by rw [← add_def, add, dif_pos hx, dif_pos hy]
+
+@[simp] lemma some_add_self_of_y_eq (hy : y₁ = W.neg_Y x₁ y₁) : some h₁ h₁' + some h₁ h₁' = 0 :=
+some_add_some_of_y_eq h₁ h₁ h₁' h₁' rfl hy
+
+@[simp] lemma some_add_some_of_y_ne (hx : x₁ = x₂) (hy : y₁ ≠ W.neg_Y x₂ y₂) :
+  some h₁ h₁' + some h₂ h₂'
+    = some (equation_add_of_eq h₁ $ Y_ne_of_Y_ne h₁ h₂ hx hy)
+      (nonsingular_add_of_eq h₁ h₁' $ Y_ne_of_Y_ne h₁ h₂ hx hy) :=
+by rw [← add_def, add, dif_pos hx, dif_neg hy]
+
+lemma some_add_some_of_y_ne' (hx : x₁ = x₂) (hy : y₁ ≠ W.neg_Y x₂ y₂) :
+  some h₁ h₁' + some h₂ h₂'
+    = -some (equation_add_of_eq' h₁ $ Y_ne_of_Y_ne h₁ h₂ hx hy)
+      (nonsingular_add_of_eq' h₁ h₁' $ Y_ne_of_Y_ne h₁ h₂ hx hy) :=
+some_add_some_of_y_ne h₁ h₂ h₁' h₂' hx hy
+
+@[simp] lemma some_add_self_of_y_ne :
```

Comment:
It is a bit confusing that the `hy` that is used here was defined a long way up in the file, contrary to the assumptions used in the preceing lemmas.

Reply:
Perhaps it will be better if I put the assumptions in the actual lemmas rather than doing `include` and `omit`?